File: windows/Tests/TestRunner.cs
namespace BarelyReal.Tests;

internal static class TestRunner
{
    private static int _passed;
    private static readonly List<(string Name, string Message)> _failed = new();

    public static void Run(string name, Action body)
    {
        try
        {
            body();
            _passed++;
            Console.WriteLine($"  ✓ {name}");
        }
        catch (Exception ex)
        {
            _failed.Add((name, ex.Message));
            Console.WriteLine($"  ✗ {name} — {ex.Message}");
        }
    }

    public static int Report()
    {
        Console.WriteLine();
        Console.WriteLine($"Passed: {_passed}  Failed: {_failed.Count}");
        return _failed.Count == 0 ? 0 : 1;
    }
}

internal sealed class AssertionException : Exception
{
    public AssertionException(string message) : base(message) { }
}

internal static class Expect
{
    public static void True(bool condition, string message)
    {
        if (!condition) throw new AssertionException(message);
    }

    public static void Equal<T>(T actual, T expected)
    {
        if (!EqualityComparer<T>.Default.Equals(actual, expected))
            throw new AssertionException($"expected {expected}, got {actual}");
    }

    public static void EqualBytes(byte[] actual, byte[] expected)
    {
        if (!actual.AsSpan().SequenceEqual(expected.AsSpan()))
            throw new AssertionException(
                $"byte mismatch: expected [{Convert.ToHexString(expected)}] got [{Convert.ToHexString(actual)}]");
    }

    public static void Throws<TException>(Action body) where TException : Exception
    {
        try
        {
            body();
        }
        catch (TException)
        {
            return;
        }
        catch (Exception other)
        {
            throw new AssertionException($"expected {typeof(TException).Name}, got {other.GetType().Name}: {other.Message}");
        }
        throw new AssertionException($"expected {typeof(TException).Name}, nothing thrown");
    }
}

File: mac/Tests/TestRunner.swift
import Foundation

enum TestFailure: Error {
    case message(String)
}

@MainActor
final class TestRunner {
    static var shared = TestRunner()

    private(set) var passed = 0
    private(set) var failed: [(String, String)] = []

    func run(_ name: String, _ block: () throws -> Void) {
        do {
            try block()
            passed += 1
            print("  ✓ \(name)")
        } catch let TestFailure.message(msg) {
            failed.append((name, msg))
            print("  ✗ \(name) — \(msg)")
        } catch {
            failed.append((name, "threw \(error)"))
            print("  ✗ \(name) — threw \(error)")
        }
    }

    func report() -> Int32 {
        print("")
        print("Passed: \(passed)  Failed: \(failed.count)")
        return failed.isEmpty ? 0 : 1
    }
}

func expect(_ condition: @autoclosure () -> Bool, _ message: String, file: StaticString = #file, line: UInt = #line) throws {
    if !condition() {
        throw TestFailure.message("\(message) [\(file):\(line)]")
    }
}

func expectEqual<T: Equatable>(_ a: T, _ b: T, file: StaticString = #file, line: UInt = #line) throws {
    if a != b {
        throw TestFailure.message("expected \(b), got \(a) [\(file):\(line)]")
    }
}

func expectThrows<E: Error & Equatable>(_ expected: E, _ block: () throws -> Void, file: StaticString = #file, line: UInt = #line) throws {
    do {
        try block()
        throw TestFailure.message("expected throw \(expected), nothing thrown [\(file):\(line)]")
    } catch let err as E {
        if err != expected {
            throw TestFailure.message("expected \(expected), got \(err) [\(file):\(line)]")
        }
    } catch let other as TestFailure {
        throw other
    } catch {
        throw TestFailure.message("expected \(expected), got \(error) [\(file):\(line)]")
    }
}
